File: backend/nadle_backend/models/logging.py
"""
로깅 시스템 모델

통합 로깅 시스템을 위한 MongoDB 모델 정의
내부 애플리케이션 로그와 외부 인프라 로그를 통합 관리
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from enum import Enum
from beanie import Document, Indexed
from pydantic import BaseModel, Field


class LogLevel(str, Enum):
    """로그 레벨 열거형"""
    ERROR = "ERROR"
    WARN = "WARN"
    INFO = "INFO"
    DEBUG = "DEBUG"


class LogSource(str, Enum):
    """로그 소스 열거형"""
    INTERNAL = "internal"  # 내부 애플리케이션
    EXTERNAL = "external"  # 외부 인프라


class ServiceType(str, Enum):
    """서비스 타입 열거형"""
    API = "api"           # FastAPI 백엔드
    WEB = "web"           # Remix 프론트엔드
    CLOUD_RUN = "cloud-run"  # Google Cloud Run
    DATABASE = "database"  # MongoDB Atlas
    REDIS = "redis"       # Redis 캐시
    VERCEL = "vercel"     # Vercel 배포


class LogContext(BaseModel):
    """로그 컨텍스트 정보"""
    user_id: Optional[str] = None
    endpoint: Optional[str] = None
    method: Optional[str] = None
    status_code: Optional[int] = None
    response_time: Optional[int] = None  # 밀리초
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    session_id: Optional[str] = None
    request_id: Optional[str] = None
    
    # 외부 인프라 관련
    infrastructure: Optional[str] = None
    instance_id: Optional[str] = None
    region: Optional[str] = None
    version: Optional[str] = None


class LogMetadata(BaseModel):
    """로그 메타데이터"""
    tags: List[str] = Field(default_factory=list)
    severity: Optional[str] = None
    error_code: Optional[str] = None
    correlation_id: Optional[str] = None
    
    # 성능 메트릭
    memory_usage: Optional[float] = None
    cpu_usage: Optional[float] = None
    disk_usage: Optional[float] = None
    
    # 외부 인프라 관련
    cloud_trace_id: Optional[str] = None
    atlas_cluster: Optional[str] = None
    vercel_deployment_id: Optional[str] = None


class LogEntry(Document):
    """
    통합 로그 엔트리 모델
    
    내부/외부 모든 로그를 저장하는 메인 모델
    """
    
    # 기본 필드
    timestamp: Indexed(datetime) = Field(default_factory=datetime.utcnow)
    level: Indexed(LogLevel)
    service: Indexed(ServiceType)
    source: Indexed(LogSource)
    message: str
    
    # 상세 정보
    context: Optional[LogContext] = None
    metadata: Optional[LogMetadata] = None
    stack_trace: Optional[str] = None
    
    # 검색 및 분석용
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None
    
    class Settings:
        name = "log_entries"
        indexes = [
            [
                ("timestamp", -1),
                ("level", 1),
                ("service", 1)
            ],
            [
                ("service", 1),
                ("source", 1),
                ("timestamp", -1)
            ],
            [
                ("level", 1),
                ("timestamp", -1)
            ],
            [
                ("context.user_id", 1),
                ("timestamp", -1)
            ],
            [
                ("context.endpoint", 1),
                ("timestamp", -1)
            ]
        ]
        # TTL 인덱스: 30일 후 자동 삭제
        ttl_index = {
            "field": "timestamp",
            "expireAfterSeconds": 30 * 24 * 60 * 60  # 30일
        }


class LogStats(BaseModel):
    """로그 통계 정보"""
    total_count: int = 0
    error_count: int = 0
    warn_count: int = 0
    info_count: int = 0
    debug_count: int = 0
    
    # 서비스별 통계
    service_stats: Dict[str, int] = Field(default_factory=dict)
    
    # 시간 범위
    start_time: datetime
    end_time: datetime


class LogFilter(BaseModel):
    """로그 필터링 조건"""
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    levels: Optional[List[LogLevel]] = None
    services: Optional[List[ServiceType]] = None
    sources: Optional[List[LogSource]] = None
    search_query: Optional[str] = None
    user_id: Optional[str] = None
    endpoint: Optional[str] = None
    
    # 페이징
    page: int = 1
    page_size: int = 50


class ErrorGrouping(Document):
    """에러 그룹핑 정보"""
    
    # 에러 식별
    error_hash: Indexed(str)  # 에러 메시지 해시
    service: Indexed(ServiceType)
    endpoint: Optional[str] = None
    error_type: Optional[str] = None
    
    # 통계
    count: int = 1
    first_seen: datetime = Field(default_factory=datetime.utcnow)
    last_seen: datetime = Field(default_factory=datetime.utcnow)
    
    # 상세 정보
    sample_message: str
    sample_stack_trace: Optional[str] = None
    
    # 분석
    severity: Optional[str] = None
    is_resolved: bool = False
    resolution_notes: Optional[str] = None
    
    class Settings:
        name = "error_groupings"
        indexes = [
            [
                ("error_hash", 1),
                ("service", 1)
            ],
            [
                ("service", 1),
                ("count", -1),
                ("last_seen", -1)
            ],
            [
                ("last_seen", -1),
                ("is_resolved", 1)
            ]
        ]


class LogSearchIndex(BaseModel):
    """로그 검색 인덱스"""
    message_tokens: List[str] = Field(default_factory=list)
    context_tokens: List[str] = Field(default_factory=list)
    full_text: str = ""


class PerformanceMetric(Document):
    """성능 메트릭 로그"""
    
    timestamp: Indexed(datetime) = Field(default_factory=datetime.utcnow)
    service: Indexed(ServiceType)
    metric_name: str
    metric_value: float
    unit: str
    
    # 컨텍스트
    endpoint: Optional[str] = None
    instance_id: Optional[str] = None
    region: Optional[str] = None
    
    # 메타데이터
    tags: Dict[str, str] = Field(default_factory=dict)
    
    class Settings:
        name = "performance_metrics"
        indexes = [
            [
                ("timestamp", -1),
                ("service", 1),
                ("metric_name", 1)
            ],
            [
                ("service", 1),
                ("endpoint", 1),
                ("timestamp", -1)
            ]
        ]
        # TTL 인덱스: 7일 후 자동 삭제
        ttl_index = {
            "field": "timestamp",
            "expireAfterSeconds": 7 * 24 * 60 * 60  # 7일
        }


# 로그 관련 응답 모델들
class LogEntryResponse(BaseModel):
    """로그 엔트리 응답 모델"""
    id: str
    timestamp: datetime
    level: LogLevel
    service: ServiceType
    source: LogSource
    message: str
    context: Optional[LogContext] = None
    metadata: Optional[LogMetadata] = None
    stack_trace: Optional[str] = None


class LogListResponse(BaseModel):
    """로그 목록 응답 모델"""
    logs: List[LogEntryResponse]
    total_count: int
    page: int
    page_size: int
    has_next: bool
    has_prev: bool


class LogDashboardResponse(BaseModel):
    """로그 대시보드 응답 모델"""
    stats: LogStats
    recent_errors: List[ErrorGrouping]
    time_series: List[Dict[str, Any]] = Field(default_factory=list)
    top_endpoints: List[Dict[str, Any]] = Field(default_factory=list)
    performance_summary: Dict[str, Any] = Field(default_factory=dict)

File: frontend/app/types/logging.ts
/**
 * 로깅 시스템 타입 정의
 * 
 * 백엔드 로깅 시스템과 일치하는 TypeScript 타입 정의
 */

// 로그 레벨 열거형
export enum LogLevel {
  ERROR = 'ERROR',
  WARN = 'WARN',
  INFO = 'INFO',
  DEBUG = 'DEBUG'
}

// 로그 소스 열거형
export enum LogSource {
  INTERNAL = 'internal',
  EXTERNAL = 'external'
}

// 서비스 타입 열거형
export enum ServiceType {
  API = 'api',
  WEB = 'web',
  CLOUD_RUN = 'cloud-run',
  DATABASE = 'database',
  REDIS = 'redis',
  VERCEL = 'vercel'
}

// 로그 컨텍스트 타입
export interface LogContext {
  user_id?: string;
  endpoint?: string;
  method?: string;
  status_code?: number;
  response_time?: number;
  ip_address?: string;
  user_agent?: string;
  session_id?: string;
  request_id?: string;
  infrastructure?: string;
  instance_id?: string;
  region?: string;
  version?: string;
}

// 로그 메타데이터 타입
export interface LogMetadata {
  tags?: string[];
  severity?: string;
  error_code?: string;
  correlation_id?: string;
  memory_usage?: number;
  cpu_usage?: number;
  disk_usage?: number;
  cloud_trace_id?: string;
  atlas_cluster?: string;
  vercel_deployment_id?: string;
}

// 로그 엔트리 타입
export interface LogEntry {
  id: string;
  timestamp: string;
  level: LogLevel;
  service: ServiceType;
  source: LogSource;
  message: string;
  context?: LogContext;
  metadata?: LogMetadata;
  stack_trace?: string;
}

// 로그 필터 타입
export interface LogFilter {
  start_time?: string;
  end_time?: string;
  levels?: LogLevel[];
  services?: ServiceType[];
  sources?: LogSource[];
  search_query?: string;
  user_id?: string;
  endpoint?: string;
  status_codes?: number[];
  regions?: string[];
  instance_ids?: string[];
  deployment_ids?: string[];
  page?: number;
  page_size?: number;
}

// 고급 로그 필터 타입 (UI 확장용)
export interface AdvancedLogFilter extends LogFilter {
  infrastructure_types?: string[];
  performance_filters?: {
    min_response_time?: number;
    max_response_time?: number;
    min_memory_usage?: number;
    max_memory_usage?: number;
    min_cpu_usage?: number;
    max_cpu_usage?: number;
  };
  error_filters?: {
    has_stack_trace?: boolean;
    error_codes?: string[];
    correlation_ids?: string[];
  };
}

// 로그 통계 타입
export interface LogStats {
  total_count: number;
  error_count: number;
  warn_count: number;
  info_count: number;
  debug_count: number;
  service_stats: Record<string, number>;
  start_time: string;
  end_time: string;
}

// 에러 그룹핑 타입
export interface ErrorGrouping {
  id: string;
  error_hash: string;
  service: ServiceType;
  endpoint?: string;
  error_type?: string;
  count: number;
  first_seen: string;
  last_seen: string;
  sample_message: string;
  sample_stack_trace?: string;
  severity?: string;
  is_resolved: boolean;
  resolution_notes?: string;
}

// 로그 목록 응답 타입
export interface LogListResponse {
  logs: LogEntry[];
  total_count: number;
  page: number;
  page_size: number;
  has_next: boolean;
  has_prev: boolean;
}

// 시계열 데이터 타입
export interface TimeSeriesData {
  timestamp: string;
  error: number;
  warn: number;
  info: number;
  debug: number;
}

// 상위 엔드포인트 타입
export interface TopEndpoint {
  endpoint: string;
  total_requests: number;
  error_count: number;
  error_rate: number;
  avg_response_time: number;
}

// 성능 요약 타입
export interface PerformanceSummary {
  services: Record<string, {
    avg_response_time: number;
    max_response_time: number;
    request_count: number;
  }>;
  overall: {
    avg_response_time: number;
    max_response_time: number;
    total_requests: number;
  };
}

// 로그 대시보드 응답 타입
export interface LogDashboardResponse {
  stats: LogStats;
  recent_errors: ErrorGrouping[];
  time_series: TimeSeriesData[];
  top_endpoints: TopEndpoint[];
  performance_summary: PerformanceSummary;
}

// 외부 로그 수집 상태 타입
export interface ExternalLogHealth {
  collectors: Record<string, {
    status: 'healthy' | 'warning' | 'error';
    last_collection_hours_ago?: number;
    error?: string;
  }>;
  last_collection_times: Record<string, string>;
  total_logs_24h: number;
  error?: string;
}

// 로깅 시스템 전체 상태 타입
export interface LoggingSystemHealth {
  status: 'healthy' | 'warning' | 'error';
  internal_logging: {
    status: 'healthy' | 'warning' | 'error';
    logs_24h: number;
  };
  external_logging: {
    status: 'healthy' | 'warning' | 'error';
    logs_24h: number;
    collectors: Record<string, any>;
  };
  total_logs_24h: number;
  timestamp: string;
}

// API 응답 공통 타입
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: {
    message: string;
    code?: string;
    details?: any;
  };
}

// 로깅 컴포넌트 Props 타입들
export interface LogStatsCardsProps {
  stats: LogStats;
  loading?: boolean;
  className?: string;
}

export interface LogErrorTopListProps {
  errors: ErrorGrouping[];
  loading?: boolean;
  className?: string;
  maxItems?: number;
}

export interface LogFilterPanelProps {
  filter: LogFilter;
  onFilterChange: (filter: LogFilter) => void;
  loading?: boolean;
  className?: string;
}

export interface LogTableProps {
  logs: LogEntry[];
  loading?: boolean;
  className?: string;
  onLogClick?: (log: LogEntry) => void;
}

export interface LogDetailModalProps {
  log: LogEntry | null;
  isOpen: boolean;
  onClose: () => void;
  className?: string;
}

export interface LoggingDashboardProps {
  autoRefresh?: boolean;
  refreshInterval?: number;
  initialLoading?: boolean;
  className?: string;
}

// 로그 레벨별 색상 및 아이콘 매핑
export const LOG_LEVEL_CONFIG = {
  [LogLevel.ERROR]: {
    color: 'red',
    bgColor: 'bg-red-50',
    textColor: 'text-red-800',
    borderColor: 'border-red-200',
    icon: '🔴',
    label: 'ERROR'
  },
  [LogLevel.WARN]: {
    color: 'yellow',
    bgColor: 'bg-yellow-50',
    textColor: 'text-yellow-800',
    borderColor: 'border-yellow-200',
    icon: '🟡',
    label: 'WARN'
  },
  [LogLevel.INFO]: {
    color: 'blue',
    bgColor: 'bg-blue-50',
    textColor: 'text-blue-800',
    borderColor: 'border-blue-200',
    icon: '🔵',
    label: 'INFO'
  },
  [LogLevel.DEBUG]: {
    color: 'gray',
    bgColor: 'bg-gray-50',
    textColor: 'text-gray-800',
    borderColor: 'border-gray-200',
    icon: '⚪',
    label: 'DEBUG'
  }
} as const;

// 서비스 타입별 색상 및 아이콘 매핑
export const SERVICE_TYPE_CONFIG = {
  [ServiceType.API]: {
    color: 'blue',
    bgColor: 'bg-blue-50',
    textColor: 'text-blue-800',
    borderColor: 'border-blue-200',
    icon: '🔧',
    label: 'API'
  },
  [ServiceType.WEB]: {
    color: 'green',
    bgColor: 'bg-green-50',
    textColor: 'text-green-800',
    borderColor: 'border-green-200',
    icon: '🌐',
    label: 'Web'
  },
  [ServiceType.CLOUD_RUN]: {
    color: 'purple',
    bgColor: 'bg-purple-50',
    textColor: 'text-purple-800',
    borderColor: 'border-purple-200',
    icon: '☁️',
    label: 'Cloud Run'
  },
  [ServiceType.DATABASE]: {
    color: 'indigo',
    bgColor: 'bg-indigo-50',
    textColor: 'text-indigo-800',
    borderColor: 'border-indigo-200',
    icon: '🗄️',
    label: 'Database'
  },
  [ServiceType.REDIS]: {
    color: 'red',
    bgColor: 'bg-red-50',
    textColor: 'text-red-800',
    borderColor: 'border-red-200',
    icon: '⚡',
    label: 'Redis'
  },
  [ServiceType.VERCEL]: {
    color: 'gray',
    bgColor: 'bg-gray-50',
    textColor: 'text-gray-800',
    borderColor: 'border-gray-200',
    icon: '🚀',
    label: 'Vercel'
  }
} as const;

// 서비스 타입 옵션 (필터용)
export const SERVICE_OPTIONS = [
  { value: ServiceType.API, label: 'API', icon: '🔧' },
  { value: ServiceType.WEB, label: 'Web', icon: '🌐' },
  { value: ServiceType.CLOUD_RUN, label: 'Cloud Run', icon: '☁️' },
  { value: ServiceType.DATABASE, label: 'Database', icon: '🗄️' },
  { value: ServiceType.REDIS, label: 'Redis', icon: '⚡' },
  { value: ServiceType.VERCEL, label: 'Vercel', icon: '🚀' }
];

// 로그 소스 옵션 (필터용)
export const SOURCE_OPTIONS = [
  { value: LogSource.INTERNAL, label: 'Internal', icon: '🏠' },
  { value: LogSource.EXTERNAL, label: 'External', icon: '🌍' }
];

// 지역 옵션 (필터용)
export const REGION_OPTIONS = [
  { value: 'asia-northeast3', label: 'Asia Northeast 3 (Seoul)', icon: '🇰🇷' },
  { value: 'asia-northeast1', label: 'Asia Northeast 1 (Tokyo)', icon: '🇯🇵' },
  { value: 'us-central1', label: 'US Central 1', icon: '🇺🇸' },
  { value: 'europe-west1', label: 'Europe West 1', icon: '🇪🇺' },
  { value: 'unknown', label: 'Unknown', icon: '❓' }
];

// 인프라 타입 옵션 (필터용)
export const INFRASTRUCTURE_OPTIONS = [
  { value: 'gcp', label: 'Google Cloud Platform', icon: '☁️' },
  { value: 'vercel', label: 'Vercel', icon: '🚀' },
  { value: 'upstash', label: 'Upstash', icon: '⚡' },
  { value: 'atlas', label: 'MongoDB Atlas', icon: '🗄️' },
  { value: 'local', label: 'Local', icon: '🏠' }
];

// 로그 소스별 색상 및 아이콘 매핑
export const LOG_SOURCE_CONFIG = {
  [LogSource.INTERNAL]: {
    color: 'blue',
    bgColor: 'bg-blue-50',
    textColor: 'text-blue-800',
    borderColor: 'border-blue-200',
    icon: '🏠',
    label: 'Internal'
  },
  [LogSource.EXTERNAL]: {
    color: 'orange',
    bgColor: 'bg-orange-50',
    textColor: 'text-orange-800',
    borderColor: 'border-orange-200',
    icon: '🌍',
    label: 'External'
  }
} as const;

// 유틸리티 함수 타입
export interface LogUtils {
  formatTimestamp: (timestamp: string) => string;
  formatDuration: (ms: number) => string;
  formatFileSize: (bytes: number) => string;
  getLogLevelConfig: (level: LogLevel) => typeof LOG_LEVEL_CONFIG[LogLevel];
  getServiceTypeConfig: (service: ServiceType) => typeof SERVICE_TYPE_CONFIG[ServiceType];
  getLogSourceConfig: (source: LogSource) => typeof LOG_SOURCE_CONFIG[LogSource];
  truncateMessage: (message: string, maxLength?: number) => string;
  highlightSearchTerm: (text: string, searchTerm: string) => string;
}

// 로그 검색 및 필터링 관련 타입
export interface LogSearchState {
  query: string;
  filters: LogFilter;
  results: LogEntry[];
  totalCount: number;
  loading: boolean;
  error: string | null;
  hasMore: boolean;
}

export interface LogSearchActions {
  setQuery: (query: string) => void;
  setFilters: (filters: Partial<LogFilter>) => void;
  search: () => Promise<void>;
  loadMore: () => Promise<void>;
  reset: () => void;
}

// 로그 대시보드 상태 타입
export interface LogDashboardState {
  data: LogDashboardResponse | null;
  loading: boolean;
  error: string | null;
  lastUpdated: string | null;
  autoRefresh: boolean;
  refreshInterval: number;
}

export interface LogDashboardActions {
  fetchDashboard: (hours?: number) => Promise<void>;
  setAutoRefresh: (enabled: boolean) => void;
  setRefreshInterval: (interval: number) => void;
  refresh: () => Promise<void>;
}

// 외부 로그 수집 관련 타입
export interface ExternalLogCollectionState {
  status: 'idle' | 'collecting' | 'success' | 'error';
  progress: number;
  message: string;
  results: Record<string, number>;
  error: string | null;
}

export interface ExternalLogCollectionActions {
  collectLogs: (hours?: number) => Promise<void>;
  checkHealth: () => Promise<void>;
  reset: () => void;
}

// 로그 정리 관련 타입
export interface LogCleanupState {
  status: 'idle' | 'cleaning' | 'success' | 'error';
  progress: number;
  deletedCount: number;
  error: string | null;
}

export interface LogCleanupActions {
  cleanup: (days: number) => Promise<void>;
  reset: () => void;
}